### PR TITLE
Try to reduce GC produced while writing headers

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

bbc34d0eda38efb4a6364561e9ac1d2319682694 introduced correct handling of "in process" setup of streams but there is some room for improvements. Often the writeHeaders(...) is completed directly which means there is not need to create the extra listener object.

Modifications:

- Only create the listener if we really need too.

Result:

Less GC

### CODE REVIEW

**Comment by normanmaurer** [comment]
@millems  @njhill PTAL... this is a followup to reduce some GC.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java [hunk 1]
```diff
@@ -420,26 +420,31 @@ private void writeHeadersFrame(
...
-                    }
-                }
-            });
-
             encoder().writeHeaders(ctx, streamId, headersFrame.headers(), headersFrame.padding(),
                     headersFrame.isEndStream(), promise);
 
             if (!promise.isDone()) {
                 numBufferedStreams++;
                 promise.addListener(bufferedStreamsListener);
```
Comment (njhill) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java]: Can also get rid of `bufferedStreamsListener` now, and just add `numBufferedStreams--` to the new listener

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java]: Done

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Pull request merged**